base/MAPL_Generic.F90

```diff
@@ -4916,7 +4941,247 @@ subroutine MAPL_TerminateImportAll ( GC, ALL, RC )
+    integer,              optional, intent(  OUT) :: RC     ! Error code:
+    !EOPI
+
+    character(len=ESMF_MAXSTR), parameter :: IAm="MAPL_ServiceAddConnection"
+    integer                               :: STATUS
+
+    type (MAPL_Connectivity), pointer     :: conn
+
+    call MAPL_ConnectivityGet(gc, connectivityPtr=conn, RC=status)
+    _VERIFY(STATUS)
+
+    call MAPL_VarServiceConnectionCreate(CONN%ServiceConnection, &
+         PROVIDER=PROVIDER, &
+         SUBSCRIBER=SUBSCRIBER, &
+         SERVICE=SERVICE, &
+         RC=STATUS  )
+    _VERIFY(STATUS)
+
+    _RETURN(ESMF_SUCCESS)
+
+  end subroutine MAPL_ServiceAddConnection
+
+  subroutine MAPL_ServiceAddProvider(GC, SERVICE, BUNDLE, RC)
+    type(ESMF_GridComp),            intent(INOUT) :: GC ! Gridded component
+    character (len=*),              intent(IN   ) :: SERVICE
+    character (len=*),              intent(IN   ) :: BUNDLE
+    integer,              optional, intent(  OUT) :: RC     ! Error code:
+
+    integer :: status
+    type(MAPL_VarServiceProviderPtr), pointer :: provider_list(:) => null()
+
+    
+    call MAPL_ServiceProviderGet(gc, provider=provider_list, RC=status)
+    _VERIFY(STATUS)
 
+    call MAPL_VarServiceProviderListCreate(provider_list, &
+         SERVICE=SERVICE, &
+         BUNDLE=BUNDLE, &
+         RC=STATUS  )
+    _VERIFY(STATUS)
+
+    _RETURN(ESMF_SUCCESS)
+  end subroutine MAPL_ServiceAddProvider
+  
+  subroutine MAPL_ServiceAddSubscriber(GC, SERVICE, VARS, RC)
+    type(ESMF_GridComp),            intent(INOUT) :: GC ! Gridded component
+    character (len=*),              intent(IN   ) :: SERVICE
+    character (len=*),              intent(IN   ) :: VARS(:)
+    integer,              optional, intent(  OUT) :: RC     ! Error code:
+
+    integer :: status
+    type(MAPL_VarServiceSubscriberPtr), pointer :: subscriber_list(:) => null()
+    
+    call MAPL_ServiceSubscriberGet(gc, subscriber=subscriber_list, RC=status)
+    _VERIFY(STATUS)
+
+    call MAPL_VarServiceSubscriberListCreate(subscriber_list, &
+         SERVICE=SERVICE, &
+         VARS = VARS, &
+         RC=STATUS  )
+    _VERIFY(STATUS)
+    
+    _RETURN(ESMF_SUCCESS)
+
+  end subroutine MAPL_ServiceAddSubscriber
+
+  subroutine MAPL_ServiceProviderGet(GC, PROVIDER, RC)
+    type(ESMF_GridComp),            intent(INOUT) :: GC ! Gridded component
+    type(MAPL_VarServiceProviderPtr), pointer :: provider(:)
+    integer,              optional, intent(  OUT) :: RC     ! Error code:
+
+    integer :: status
+    type (MAPL_MetaComp), pointer     :: MAPLOBJ
+
+    !get MAPL
+    call MAPL_InternalStateRetrieve ( GC, MAPLOBJ, RC=STATUS )
+    _VERIFY(STATUS)
+
+    provider => maplobj%provider_list
+    _RETURN(ESMF_SUCCESS)
+    
+  end subroutine MAPL_ServiceProviderGet
+
+  subroutine MAPL_ServiceSubscriberGet(GC, SUBSCRIBER, RC)
+    type(ESMF_GridComp),            intent(INOUT) :: GC ! Gridded component
+    type(MAPL_VarServiceSubscriberPtr), pointer :: subscriber(:)
```

Comment:
"subscriber" => "subscribers"  (plural)
The procedure name should also be plural.